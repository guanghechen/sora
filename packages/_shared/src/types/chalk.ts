import type { ColorSupportLevelEnum } from '../constant'

export interface IChalkColorSupport {
  /**
   * The color level.
   */
  level: ColorSupportLevelEnum

  /**
   * Whether basic 16 colors are supported.
   */
  hasBasic: boolean

  /**
   * Whether ANSI 256 colors are supported.
   */
  has256: boolean

  /**
   * Whether Truecolor 16 million colors are supported.
   */
  has16m: boolean
}

export interface IColorPair {
  /**
   * The ANSI terminal control sequence for starting this style.
   */
  readonly open: string

  /**
   * The ANSI terminal control sequence for ending this style.
   */
  readonly close: string
}

export interface IModifier {
  /**
   * Resets the current color chain.
   */
  readonly reset: IColorPair

  /**
   * Make text bold.
   */
  readonly bold: IColorPair

  /**
   * Emitting only a small amount of light.
   */
  readonly dim: IColorPair

  /**
   * Make text italic. (Not widely supported)
   */
  readonly italic: IColorPair

  /**
   * Make text underline. (Not widely supported)
   */
  readonly underline: IColorPair

  /**
   * Make text overline.
   * Supported on VTE-based terminals, the GNOME terminal, mintty, and Git Bash.
   */
  readonly overline: IColorPair

  /**
   * Inverse background and foreground colors.
   */
  readonly inverse: IColorPair

  /**
   * Prints the text, but makes it invisible.
   */
  readonly hidden: IColorPair

  /**
   * Puts a horizontal line through the center of the text. (Not widely supported)
   */
  readonly strikethrough: IColorPair
}

export interface IChalkColor {
  readonly close: string
  ansi(code: number): string
  ansi256(code: number): string
  ansi16m(r: number, g: number, b: number): string
}

export interface IForegroundColor extends IChalkColor {
  readonly black: IColorPair
  readonly red: IColorPair
  readonly green: IColorPair
  readonly yellow: IColorPair
  readonly blue: IColorPair
  readonly cyan: IColorPair
  readonly magenta: IColorPair
  readonly white: IColorPair

  // Alias for `blackBright`.
  readonly gray: IColorPair

  // Alias for `blackBright`.
  readonly grey: IColorPair

  readonly blackBright: IColorPair
  readonly redBright: IColorPair
  readonly greenBright: IColorPair
  readonly yellowBright: IColorPair
  readonly blueBright: IColorPair
  readonly cyanBright: IColorPair
  readonly magentaBright: IColorPair
  readonly whiteBright: IColorPair
}

export interface IBackgroundColor extends IChalkColor {
  readonly bgBlack: IColorPair
  readonly bgRed: IColorPair
  readonly bgGreen: IColorPair
  readonly bgYellow: IColorPair
  readonly bgBlue: IColorPair
  readonly bgCyan: IColorPair
  readonly bgMagenta: IColorPair
  readonly bgWhite: IColorPair

  // Alias for `bgBlackBright`.
  readonly bgGray: IColorPair

  // Alias for `bgBlackBright`.
  readonly bgGrey: IColorPair

  readonly bgBlackBright: IColorPair
  readonly bgRedBright: IColorPair
  readonly bgGreenBright: IColorPair
  readonly bgYellowBright: IColorPair
  readonly bgBlueBright: IColorPair
  readonly bgCyanBright: IColorPair
  readonly bgMagentaBright: IColorPair
  readonly bgWhiteBright: IColorPair
}

export interface IChalkStyler {
  readonly open: string
  readonly close: string
  readonly openAll: string
  readonly closeAll: string
  readonly parent: Readonly<IChalkStyler> | undefined
}

export type IChalkColorName = keyof IChalkColor
export type IModifyExcludeColorName = Exclude<keyof IModifier, IChalkColorName>
export type IForegroundColorExcludeColorName = Exclude<keyof IForegroundColor, IChalkColorName>
export type IBackgroundColorExcludeColorName = Exclude<keyof IBackgroundColor, IChalkColorName>
export type IStyleName =
  | IModifyExcludeColorName
  | IForegroundColorExcludeColorName
  | IBackgroundColorExcludeColorName

export interface IAnsiStyles extends IModifier, IForegroundColor, IBackgroundColor {
  modifier: IModifier
  color: IForegroundColor
  bgColor: IBackgroundColor
  codes: ReadonlyMap<number, number>
}

export interface IBasicStyles {
  readonly level: 0 | 1 | 2 | 3
  readonly styler: IChalkStyler | undefined
  readonly isEmpty: boolean
  readonly visible: IChalkBuilder
  rgb(r: number, g: number, b: number): IChalkBuilder
  hex(hex: string): IChalkBuilder
  ansi256(code: number): IChalkBuilder
  bgRgb(r: number, g: number, b: number): IChalkBuilder
  bgHex(hex: string): IChalkBuilder
  bgAnsi256(code: number): IChalkBuilder
}

export type IChalkStyles = {
  readonly [key in IStyleName]: IChalkBuilder
} & IBasicStyles

export interface IChalkBuilder extends IChalkStyles {
  // Apply colors
  (...texts: string[]): string
}

export interface IChalk extends IChalkStyles {}
