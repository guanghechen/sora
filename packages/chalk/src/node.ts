import { ColorSupportLevelEnum } from '@guanghechen/internal'
import { supportsColor, supportsColorStderr } from './node/supports-color'
import { Chalk } from '.'

export { supportsColor, supportsColorStderr } from './node/supports-color'
export const chalk = Chalk.create(
  supportsColor ? supportsColor.level : ColorSupportLevelEnum.DISABLED,
)
export const chalkStderr = Chalk.create(
  supportsColorStderr ? supportsColorStderr.level : ColorSupportLevelEnum.DISABLED,
)
