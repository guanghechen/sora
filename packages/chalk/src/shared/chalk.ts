import { ColorSupportLevelEnum } from '@guanghechen/internal'
import { Styles } from './styles'

export class Chalk extends Styles {
  private constructor(level: ColorSupportLevelEnum) {
    super(level, undefined, false)
  }

  static #instanceMap: Map<ColorSupportLevelEnum, Chalk> = new Map()
  public static create(level: ColorSupportLevelEnum): Chalk {
    if (
      !Number.isInteger(level) ||
      level < ColorSupportLevelEnum.DISABLED ||
      level > ColorSupportLevelEnum.True16m
    ) {
      throw new Error(
        `The \`level\` option should be an integer from ${ColorSupportLevelEnum.DISABLED} to ${ColorSupportLevelEnum.True16m}`,
      )
    }

    let instance: Chalk | undefined = this.#instanceMap.get(level)
    if (instance === undefined) {
      instance = new Chalk(level)
      this.#instanceMap.set(level, instance)
    }
    return instance
  }
}
