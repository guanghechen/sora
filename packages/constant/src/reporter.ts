export enum ReporterLevelEnum {
  DEBUG = 'debug',
  VERBOSE = 'verbose',
  INFO = 'info',
  WARN = 'warn',
  ERROR = 'error',
  FATAL = 'fatal',
}

export const LevelOrdinalMap: Record<ReporterLevelEnum, number> = Object.freeze({
  [ReporterLevelEnum.DEBUG]: 1,
  [ReporterLevelEnum.VERBOSE]: 2,
  [ReporterLevelEnum.INFO]: 3,
  [ReporterLevelEnum.WARN]: 4,
  [ReporterLevelEnum.ERROR]: 5,
  [ReporterLevelEnum.FATAL]: 6,
})
