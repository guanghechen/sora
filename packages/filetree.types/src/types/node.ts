import type { FileTreeNodeTypeEnum } from '../constant'

export interface IFileTreeNodeStat {
  /**
   * The type of the file tree node.
   */
  readonly type: FileTreeNodeTypeEnum

  /**
   * The creation timestamp in milliseconds elapsed since January 1, 1970 00:00:00 UTC.
   */
  readonly ctime: number

  /**
   * The modification timestamp in milliseconds elapsed since January 1, 1970 00:00:00 UTC.
   *
   * *Note:* If the file changed, it is important to provide an updated `mtime` that advanced
   * from the previous value. Otherwise there may be optimizations in place that will not show
   * the updated file contents in an editor for example.
   */
  readonly mtime: number

  /**
   * The size in bytes.
   *
   * *Note:* If the file changed, it is important to provide an updated `size`. Otherwise there
   * may be optimizations in place that will not show the updated file contents in an editor for
   * example.
   */
  readonly size: number
}

export interface IFileTreeFileNode extends IFileTreeNodeStat {
  readonly type: FileTreeNodeTypeEnum.FILE
  /**
   * File name.
   */
  readonly name: string
}

export interface IFileTreeFolderNode extends IFileTreeNodeStat {
  readonly type: FileTreeNodeTypeEnum.FOLDER
  /**
   * Folder name.
   */
  readonly name: string
  /**
   * Children nodes.
   */
  readonly children: ReadonlyArray<IFileTreeNode>
}

export type IFileTreeNode = IFileTreeFolderNode | IFileTreeFileNode
