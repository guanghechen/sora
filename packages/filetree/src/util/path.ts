const regex = /[/\\]+/g

/**
 * Split the pathFromRoot into path pieces.
 * @param path
 * @returns
 */
export function splitPathFromRoot(path: string): string[] {
  if (path.startsWith('/') || path.startsWith('\\')) {
    throw new TypeError(`[splitPathFromRoot] pathFromRoot should be a relative path (${path}).`)
  }

  const rawPieces: string[] = path.split(regex)
  const pieces: string[] = []
  for (const piece of rawPieces) {
    if (piece.length <= 0) continue
    if (piece === '.') continue
    if (piece === '..') pieces.pop()
    else pieces.push(piece)
  }
  return pieces
}
