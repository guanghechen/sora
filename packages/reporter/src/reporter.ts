import type { IChalk } from '@guanghechen/chalk.types'
import type { IReporter, IReporterFlights } from '@guanghechen/reporter.types'
import { ReporterLevelEnum } from '@guanghechen/reporter.types'
import dayjs from 'dayjs'
import { parseOptionsFromArgs } from './args'
import { normalizeString } from './format'
import type { ILevelStyleMap } from './level'

export interface IReporterOptions {
  baseName?: string
  level?: ReporterLevelEnum
  flights?: Partial<IReporterFlights>
  placeholderRegex?: RegExp
}

const write =
  typeof process !== 'undefined'
    ? (text: string): void => {
        process.stdout.write(text)
      }
    : (text: string): void => {
        console.log(text)
      }
const getDefaultArgs = (): string[] => {
  if (typeof process === 'undefined') return []
  return process.argv
}

export class Reporter implements IReporter {
  public readonly level: ReporterLevelEnum
  public readonly levelStyleMap: ILevelStyleMap
  public readonly flights: Readonly<IReporterFlights>
  public readonly placeholderRegex: RegExp
  protected _baseName: string
  protected _divisionName: string

  constructor(chalk: IChalk, options_: IReporterOptions = {}, args_: string[] = getDefaultArgs()) {
    const argOptions: IReporterOptions = parseOptionsFromArgs(args_)
    const options: IReporterOptions = {
      ...options_,
      ...argOptions,
      flights: {
        ...options_.flights,
        ...argOptions.flights,
      },
    }

    this.level = options.level ?? ReporterLevelEnum.INFO
    this.levelStyleMap = Object.freeze({
      [ReporterLevelEnum.DEBUG]: {
        title: 'debug',
        labelChalk: { fg: chalk.grey },
        contentChalk: { fg: chalk.grey },
      },
      [ReporterLevelEnum.VERBOSE]: {
        title: 'verb ',
        labelChalk: { fg: chalk.cyan },
        contentChalk: { fg: chalk.cyan },
      },
      [ReporterLevelEnum.INFO]: {
        title: 'info ',
        labelChalk: { fg: chalk.green },
        contentChalk: { fg: chalk.green },
      },
      [ReporterLevelEnum.WARN]: {
        title: 'warn ',
        labelChalk: { fg: chalk.yellow },
        contentChalk: { fg: chalk.yellow },
      },
      [ReporterLevelEnum.ERROR]: {
        title: 'error',
        labelChalk: { fg: chalk.red },
        contentChalk: { fg: chalk.red },
      },
      [ReporterLevelEnum.FATAL]: {
        title: 'fatal',
        labelChalk: { fg: chalk.black, bg: chalk.bgRed },
        contentChalk: { fg: chalk.redBright },
      },
    })

    this.flights = Object.freeze({
      date: options.flights?.date ?? false,
      title: options.flights?.title ?? true,
      inline: options.flights?.inline ?? true,
      colorful: options.flights?.colorful ?? true,
    })
    this.placeholderRegex = options.placeholderRegex ?? /(?<!\\)\{\}/g
    this._baseName = options.baseName ?? ''
    this._divisionName = ''
  }

  public get name(): string {
    if (this._baseName === '') return this._divisionName ?? ''
    if (this._divisionName === '') return this._baseName
    return this._baseName + '#' + this._divisionName
  }

  public setDivisionName(divisionName: string): void {
    this._divisionName = divisionName
  }

  public write(text: string): void {
    write(text)
  }

  public debug(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(ReporterLevelEnum.DEBUG, messageFormat, ...messages)
  }

  public verbose(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(ReporterLevelEnum.VERBOSE, messageFormat, ...messages)
  }

  public info(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(ReporterLevelEnum.INFO, messageFormat, ...messages)
  }

  public warn(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(ReporterLevelEnum.WARN, messageFormat, ...messages)
  }

  public error(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(ReporterLevelEnum.ERROR, messageFormat, ...messages)
  }

  public fatal(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(ReporterLevelEnum.FATAL, messageFormat, ...messages)
  }

  // write a log record.
  public log(
    level: ReporterLevelEnum,
    messageFormat: string | unknown,
    ...messages: unknown[]
  ): void {
    if (!level || level < this.level) return
    const header = this.formatHeader(level, new Date())

    let newline = false
    const formatPattern: string = typeof messageFormat === 'string' ? messageFormat : ''
    const items: string[] = (
      typeof messageFormat === 'string' ? messages : [messageFormat, ...messages]
    ).map(msg => {
      const text = this.formatSingleMessage(msg)
      if (text.endsWith('\n')) {
        newline = true
        return '\n' + text
      }
      return text
    })

    let unpairedIdx = 0
    let message =
      items.length > 0
        ? formatPattern.replace(this.placeholderRegex, m => {
            const value = items[unpairedIdx]
            unpairedIdx += 1
            return value === undefined ? m : value
          })
        : formatPattern
    if (unpairedIdx < items.length) message += ' ' + items.slice(unpairedIdx).join(' ')
    if (!newline && !message.endsWith('\n')) message += '\n'

    this.write(this.format(level, header, message))
  }

  // format a log record.
  protected format(level: ReporterLevelEnum, header: string, message: string): string {
    const content: string = this.formatContent(level, message)
    return header.length > 0 ? header + ' ' + content : content
  }

  // format a log record's header.
  protected formatHeader(level: ReporterLevelEnum, date: Date): string {
    const dateText: string = this.flights.date
      ? this.formatContent(level, dayjs(date).format('YYYY-MM-DD HH:mm:ss'))
      : ''

    const levelStyle = this.levelStyleMap[level]
    let levelText = levelStyle.title
    if (this.flights.colorful) {
      levelText = levelStyle.labelChalk.fg(levelText)
      if (levelStyle.labelChalk.bg) levelText = levelStyle.labelChalk.bg(levelText)
    }

    const titleText: string = this.flights.title
      ? this.formatContent(level, '[' + this.name + ']')
      : ''

    let result = ''
    if (dateText) result += dateText + ' '
    result += levelText
    if (titleText) result += ' ' + titleText
    return result
  }

  protected formatContent(level: ReporterLevelEnum, message: string): string {
    let text: string = message
    if (this.flights.colorful) {
      const levelStyle = this.levelStyleMap[level]
      text = levelStyle.contentChalk.fg(text)
      if (levelStyle.contentChalk.bg) text = levelStyle.contentChalk.bg(text)
    }
    return text
  }

  // format a log record part message according its type.
  public formatSingleMessage(message: unknown): string {
    return normalizeString(message ? message : String(message), this.flights.inline)
  }
}
