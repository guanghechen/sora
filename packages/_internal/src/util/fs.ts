import type { IReporter } from '@guanghechen/reporter.types'
import type { WriteFileOptions } from 'node:fs'
import { existsSync, mkdirSync, statSync } from 'node:fs'
import { mkdir as fsMkdir, rm as fsRm, writeFile as fsWriteFile } from 'node:fs/promises'
import { dirname as pathDirname } from 'node:path'

/**
 * Remove file/folder recursively.
 * @param fileOrDirPath
 */
export async function rm(fileOrDirPath: string): Promise<void> {
  if (existsSync(fileOrDirPath)) {
    await fsRm(fileOrDirPath, { recursive: true, force: true })
  }
}

/**
 * Remove all files under the given directory path.
 * @param dirpath
 * @param createIfNotExist
 * @param reporter
 */
export async function emptyDir(
  dirpath: string,
  createIfNotExist = true,
  reporter?: IReporter,
): Promise<void> {
  if (existsSync(dirpath)) {
    if (!statSync(dirpath).isDirectory()) {
      reporter?.warn('[emptyDir] not a directory. dirpath: {}', dirpath)
      return
    }

    // Print verbose log.
    reporter?.verbose?.('[emptyDir] emptying: {}', dirpath)

    await rm(dirpath)
    await fsMkdir(dirpath, { recursive: true })
  } else {
    if (createIfNotExist) await fsMkdir(dirpath, { recursive: true })
  }
}

/**
 * Ensure critical filepath exists.
 *
 * @param filepath
 */
export function ensureCriticalFilepathExistsSync(filepath: string | null): void | never {
  let errMsg: string | null = null

  if (filepath == null) {
    errMsg = `Invalid path: ${filepath}.`
  } else if (!existsSync(filepath)) {
    errMsg = `Not found: ${filepath}.`
  } else if (!statSync(filepath).isFile()) {
    errMsg = `Not a file: ${filepath}.`
  }

  if (errMsg !== null) throw new Error(errMsg)
}

/**
 * Create a path of directories.
 *
 * @param filepath  the give file path
 * @param isDir     Whether the given path is a directory
 * @param reporter
 */
export function mkdirsIfNotExists(filepath: string, isDir: boolean, reporter?: IReporter): void {
  const dirpath = isDir ? filepath : pathDirname(filepath)
  if (existsSync(dirpath)) return

  // Print verbose log.
  reporter?.verbose?.(`mkdirs: ${dirpath}`)

  mkdirSync(dirpath, { recursive: true })
}

/**
 * If the path is not existed, created before write.
 *
 * @param filepath
 * @param content
 * @param options
 */
export async function writeFile(
  filepath: string,
  content: string | NodeJS.ArrayBufferView,
  options?: WriteFileOptions,
): Promise<void> {
  const dirpath = pathDirname(filepath)
  await fsMkdir(dirpath, { recursive: true })
  await fsWriteFile(filepath, content, options)
}
/**
 * Check whether if the filepath is a file path. (synchronizing)
 *
 * @param filepath   file path
 */
export function isFileSync(filepath: string | null): boolean {
  if (filepath == null) return false
  if (!existsSync(filepath)) return false
  const stat = statSync(filepath)
  return stat.isFile()
}
