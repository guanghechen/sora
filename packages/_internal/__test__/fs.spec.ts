// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { chalk } from '@guanghechen/chalk/node'
import { createReporterMock } from '@guanghechen/helper-jest'
import { Reporter, ReporterLevelEnum } from '@guanghechen/reporter'
import { desensitize, locateFixtures } from 'jest.helper'
import { existsSync } from 'node:fs'
import path from 'node:path'
import {
  emptyDir,
  ensureCriticalFilepathExistsSync,
  isFileSync,
  mkdirsIfNotExists,
  rm,
  writeFile,
} from '../src'

describe('empty', () => {
  const fictitiousDir = locateFixtures('fictitiousDir_c4266163b919b')

  afterEach(async () => {
    await rm(fictitiousDir)
  })

  test('basic', async () => {
    const dirpathA = path.join(fictitiousDir, 'A')
    const dirpathB = path.join(fictitiousDir, 'B')
    const filepathA = path.join(fictitiousDir, 'a.txt')
    const filepathB = path.join(dirpathB, 'b.txt')
    const filepathC = path.join(dirpathB, 'c.txt')

    expect(existsSync(fictitiousDir)).toEqual(false)

    await emptyDir(fictitiousDir)
    expect(existsSync(fictitiousDir)).toEqual(true)

    await rm(fictitiousDir)
    expect(existsSync(fictitiousDir)).toEqual(false)

    mkdirsIfNotExists(fictitiousDir, true)
    expect(existsSync(fictitiousDir)).toEqual(true)

    expect(existsSync(dirpathA)).toEqual(false)
    expect(existsSync(dirpathB)).toEqual(false)
    expect(existsSync(filepathA)).toEqual(false)
    expect(existsSync(filepathB)).toEqual(false)
    expect(existsSync(filepathC)).toEqual(false)
    await writeFile(filepathA, 'hello, A', 'utf8')
    await writeFile(filepathB, 'hello, B', 'utf8')
    await writeFile(filepathC, 'hello, C', 'utf8')
    expect(existsSync(dirpathA)).toEqual(false)
    expect(existsSync(dirpathB)).toEqual(true)
    expect(existsSync(filepathA)).toEqual(true)
    expect(existsSync(filepathB)).toEqual(true)
    expect(existsSync(filepathC)).toEqual(true)
    expect(isFileSync(filepathA)).toEqual(true)
    expect(isFileSync(filepathB)).toEqual(true)
    expect(isFileSync(filepathC)).toEqual(true)
    expect(isFileSync(dirpathA)).toEqual(false)
    expect(isFileSync(dirpathB)).toEqual(false)

    mkdirsIfNotExists(dirpathA, true)
    expect(existsSync(dirpathA)).toEqual(true)
    expect(isFileSync(filepathA)).toEqual(true)
    expect(isFileSync(filepathB)).toEqual(true)
    expect(isFileSync(filepathC)).toEqual(true)
    expect(isFileSync(dirpathA)).toEqual(false)
    expect(isFileSync(dirpathB)).toEqual(false)

    await emptyDir(fictitiousDir, false)
    expect(existsSync(fictitiousDir)).toEqual(true)
    expect(existsSync(dirpathA)).toEqual(false)
    expect(existsSync(dirpathB)).toEqual(false)
    expect(existsSync(filepathA)).toEqual(false)
    expect(existsSync(filepathB)).toEqual(false)
    expect(existsSync(filepathC)).toEqual(false)
    expect(isFileSync(filepathA)).toEqual(false)
    expect(isFileSync(filepathB)).toEqual(false)
    expect(isFileSync(filepathC)).toEqual(false)
    expect(isFileSync(dirpathA)).toEqual(false)
    expect(isFileSync(dirpathB)).toEqual(false)
  })

  test('reporter', async () => {
    const reporter = new Reporter(chalk, {
      level: ReporterLevelEnum.VERBOSE,
      flights: {
        colorful: false,
      },
    })
    const loggerMock = createReporterMock({ reporter, desensitize })

    await emptyDir(fictitiousDir, undefined, reporter)
    await emptyDir(fictitiousDir, false, reporter)
    await emptyDir(fictitiousDir, true, reporter)

    expect(loggerMock.getIndiscriminateAll()).toMatchSnapshot()
    loggerMock.restore()
  })
})

describe('ensureCriticalFilepathExistsSync', () => {
  test('null / undefined', () => {
    expect(() => void ensureCriticalFilepathExistsSync(null)).toThrow('Invalid path: null.')
  })

  test('not found', () => {
    expect(
      () =>
        void ensureCriticalFilepathExistsSync(locateFixtures('basic/config.json-non-existed---22')),
    ).toThrow('Not found:')
  })

  test('not a file', () => {
    expect(() => void ensureCriticalFilepathExistsSync(locateFixtures('basic/'))).toThrow(
      'Not a file:',
    )
  })

  test('valid', () => {
    expect(
      () => void ensureCriticalFilepathExistsSync(locateFixtures('basic/config.yml')),
    ).not.toThrow()
  })
})

describe('mkdirsIfNotExists', () => {
  test('directory existed', () => {
    const dirpath = locateFixtures('basic')
    expect(existsSync(dirpath)).toBe(true)
    mkdirsIfNotExists(dirpath, true)
    expect(existsSync(dirpath)).toBe(true)
  })

  test('filepath', () => {
    const dirpath = locateFixtures('basic/config.yml')
    expect(existsSync(dirpath)).toBe(true)
    mkdirsIfNotExists(dirpath, false)
    expect(existsSync(dirpath)).toBe(true)
  })

  test('mkdirs', async () => {
    const dirpath = locateFixtures('basic--non-existed')
    expect(existsSync(dirpath)).toBe(false)
    mkdirsIfNotExists(dirpath, true)
    expect(existsSync(dirpath)).toBe(true)
    await rm(dirpath)
  })

  test('mkdirs (with reporter)', async () => {
    const reporter = new Reporter(chalk, {
      level: ReporterLevelEnum.VERBOSE,
      flights: {
        colorful: false,
      },
    })
    const reporterMock = createReporterMock({ reporter, desensitize })

    const dirpath = locateFixtures('basic--non-existed--2')
    expect(existsSync(dirpath)).toBe(false)
    mkdirsIfNotExists(dirpath, true, reporter)
    expect(existsSync(dirpath)).toBe(true)

    await rm(dirpath)

    expect(reporterMock.getIndiscriminateAll()).toMatchSnapshot()
    reporterMock.restore()
  })
})
