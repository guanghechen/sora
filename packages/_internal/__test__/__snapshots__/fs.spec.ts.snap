// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty reporter 1`] = `
[
  [
    "verb  [] [emptyDir] emptying: <$WORKSPACE$>/packages/_internal/__test__/fixtures/fictitiousDir_c4266163b919b
",
  ],
  [
    "verb  [] [emptyDir] emptying: <$WORKSPACE$>/packages/_internal/__test__/fixtures/fictitiousDir_c4266163b919b
",
  ],
]
`;

exports[`mkdirsIfNotExists mkdirs (with reporter) 1`] = `
[
  [
    "verb  [] mkdirs: <$WORKSPACE$>/packages/_internal/__test__/fixtures/basic--non-existed--2
",
  ],
]
`;
